### ISSUE TITLE
Google Cloud deserialisation needs better error handling

### ISSUE THREAD

**Issue by francisdb** [akka/alpakka#834]
I'm seeing this when trying to publish a message
_Notice that the stack trace does not even contain alpakka_

(this specific one is for a topic that does not exist)

```
spray.json.DeserializationException: Object is missing required member 'messageIds'
	at spray.json.package$.deserializationError(package.scala:23)
	at spray.json.ProductFormats.fromField(ProductFormats.scala:60)
	at spray.json.ProductFormats.fromField$(ProductFormats.scala:50)
	at spray.json.DefaultJsonProtocol$.fromField(DefaultJsonProtocol.scala:30)
	at spray.json.ProductFormatsInstances$$anon$1.read(ProductFormatsInstances.scala:34)
	at spray.json.ProductFormatsInstances$$anon$1.read(ProductFormatsInstances.scala:26)
	at akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport.$anonfun$sprayJsonUnmarshaller$1(SprayJsonSupport.scala:30)
	at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
	at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
	at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.util.NoSuchElementException: key not found: messageIds
	at scala.collection.immutable.Map$Map1.apply(Map.scala:108)
	at spray.json.ProductFormats.fromField(ProductFormats.scala:57)
	... 20 common frames omitted
```

```
spray.json.DeserializationException: Object is missing required member 'access_token'
    at spray.json.package$.deserializationError(package.scala:23)
    at spray.json.ProductFormats.fromField(ProductFormats.scala:60)
    at spray.json.ProductFormats.fromField$(ProductFormats.scala:50)
    at spray.json.DefaultJsonProtocol$.fromField(DefaultJsonProtocol.scala:30)
    at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:78)
    at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:68)
    at akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport.$anonfun$sprayJsonUnmarshaller$1(SprayJsonSupport.scala:30)
    at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
    at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
    at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)
    at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
    at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
    at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
    at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
    at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
    at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

java.util.NoSuchElementException: key not found: access_token
    at scala.collection.immutable.Map$Map2.apply(Map.scala:129)
    at spray.json.ProductFormats.fromField(ProductFormats.scala:57)
    at spray.json.ProductFormats.fromField$(ProductFormats.scala:50)
    at spray.json.DefaultJsonProtocol$.fromField(DefaultJsonProtocol.scala:30)
    at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:78)
    at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:68)
    at akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport.$anonfun$sprayJsonUnmarshaller$1(SprayJsonSupport.scala:30)
```

**Comment by ruioliveiras** [comment]
+1

**Comment by damianbl** [comment]
I also have the issue. Any updates?

**Comment by 2m** [comment]
The error comes from the unmarshalling that is done by the akka-http which is used to create requests and parse responses. @damianbl would you be willing to try out the [new Pub/Sub connector that uses gRPC](https://github.com/akka/alpakka/pull/1162)? It has all of the model classes already generated so there should not be any problems regarding parsing the responses. If so, I could publish a snapshot version of the new Pub/Sub connector for you to try out.

**Comment by damianbl** [comment]
I have already debugged it. This issue is present when the response is unexpected. For example when the subscription/topic is not found (404). Then it cannot find the `messageIds` in that response.

So when you see this error make sure that your setup is correct.

Anyway, this exceptional case should be handled better I think. I will try to create a PR for this.

**Comment by 2m** [comment]
A PR would be awesome!

**Comment by damianbl** [comment]
https://github.com/akka/alpakka/pull/1228

**Comment by l15k4** [comment]
Guys this happens in alpakka google cloud storage too, it might be the same problem when I look at the code : 
```
Wrapped by: spray.json.DeserializationException: Object is missing required member 'access_token'
	at spray.json.package$.deserializationError(package.scala:23)
	at spray.json.ProductFormats.fromField(ProductFormats.scala:61)
	at spray.json.ProductFormats.fromField$(ProductFormats.scala:51)
	at spray.json.DefaultJsonProtocol$.fromField(DefaultJsonProtocol.scala:30)
	at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:81)
	at spray.json.ProductFormatsInstances$$anon$3.read(ProductFormatsInstances.scala:71)
	at akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport.$anonfun$sprayJsonUnmarshaller$1(SprayJsonSupport.scala:30)
	at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
	at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
	at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:45)
	at akka.http.scaladsl.util.FastFuture$.map$extension(FastFuture.scala:23)
	at akka.http.scaladsl.unmarshalling.Unmarshaller.$anonfun$map$3(Unmarshaller.scala:26)
	at akka.http.scaladsl.unmarshalling.Unmarshaller.$anonfun$transform$3(Unmarshaller.scala:23)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshal.to(Unmarshal.scala:25)
	at akka.stream.alpakka.googlecloud.storage.impl.GoogleTokenApi.$anonfun$getAccessToken$1(GoogleTokenApi.scala:54)
```

**Comment by francisdb** [comment]
See also #773

**Comment by ennru** [comment]
As far as I can see these have been addressed in #1228 and #2164 

Closing.
